### PR TITLE
Reimplement `TransactAsset` with `FungibleLedger` and `AssetConfig` traits

### LINKED ISSUE #534: Simplify `AssetTransactor` logic with the `ConcreteFungibleLedger` trait
Right now the `AssetTransactor` uses a tuple with 2 implementations of the transact functionality, called one after the other during exection.
This can be simplified with the unified interface of fungible and fungibles in `ConcreteFungiblesLedger` trait.

### PR DESCRIPTION
**Pull request by ghzlatarev**
## Description

closes: #534

* Replace `FungiblesTransactor` and `LocalAssetTransactor` tuple with our own `AssetTransactor` implementor called `MultiAssetAdapter`
* `MultiAssetAdapter` uses the unified interface of `FungibleLedger` trait to implement only the `withdraw_asset` and `deposit_asset` functions from `TransactAsset`
* The rest of the `TransactAsset` interface like can_check_in, teleport, etc are not required. Even though they were implemented in our old configuration they were not used.
* An additional `burn` function and implementation had to be added to our `FungiblesLedger` trait. This implementation allows accounts to go down to 0, as that was the behavior of the old configuration.
* `ReducedToZero` return results in `can_withdraw` implementation were considered an error, when in fact they are a success case, so switched that.
* XCM tests were refactored to account for the new configuration
* Refactored 1 manta-pay test which was supposed to fail because of attempting to mint more private assets than the available public assets of an account, but was actually minting the exact available amount. However since `ReducedToZero` was considered an error, this case was also holding up.
---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests.
- [x] Updated relevant documentation in the code.
- [x] Added **one** line describing your change in `<branch>/CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [x] If runtime changes, need to update the version numbers properly:
   * `authoring_version`: The version of the authorship interface. An authoring node will not attempt to author blocks unless this is equal to its native runtime.
   * `spec_version`: The version of the runtime specification. A full node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of spec_name, spec_version, and authoring_version are the same between Wasm and native.
   * `impl_version`: The version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the actual code may be different, it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are about the only changes that could be made which would result in only the impl_version changing.
   * `transaction_version`: The version of the extrinsics interface. This number must be updated in the following circumstances: extrinsic parameters (number, order, or types) have been changed; extrinsics or pallets have been removed; or the pallet order in the construct_runtime! macro or extrinsic order in a pallet has been changed. You can run the `metadata_diff.yml` workflow for help. If this number is updated, then the `spec_version` must also be updated 
- [x] Verify benchmarks & weights have been updated for any modified runtime logics
- [x] If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- [x] If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- [x] If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

### CODE REVIEW

File: primitives/src/assets.rs
Comment (ghzlatarev) [inline primitives/src/assets.rs:92]: clean up

File: primitives/src/assets.rs
Comment (ghzlatarev) [inline primitives/src/assets.rs]: `AllowDeath` because the `can_withdraw` impl uses `keep_alive = false` argument

File: primitives/src/assets.rs
Comment (ghzlatarev) [inline primitives/src/assets.rs]: `ReducedToZero` is actually a success case, because it happens when the account doesn't have to be kept alive.

File: primitives/src/xcm.rs
Comment (ghzlatarev) [inline primitives/src/xcm.rs:416]: import xcm::v2::Error

File: runtime/calamari/src/assets_config.rs
Comment (ghzlatarev) [inline runtime/calamari/src/assets_config.rs:126]: remove commented code.

File: runtime/calamari/src/xcm_config.rs
Comment (ghzlatarev) [inline runtime/calamari/src/xcm_config.rs:136]: Documentation. Copy over to all other files.

File: runtime/common/tests/xcm_mock/mod.rs
Comment (ghzlatarev) [inline runtime/common/tests/xcm_mock/mod.rs:110]: This is kind-of hidden detail of the tests but it's helpful because it sets the native asset and moves the next asset id along, so the `AssetConfig` impl can be used directly in the tests.

File: runtime/common/tests/xcm_mock/parachain.rs
Comment (ghzlatarev) [inline runtime/common/tests/xcm_mock/parachain.rs:202]: Copy over documentation.

File: primitives/src/assets.rs
Comment (ghzlatarev) [inline primitives/src/assets.rs]: The `transfer()` impl actually tries to keep accounts alive, instead of allowing death. 
This is inconsistent with the behavior of `TransactAsset::withdraw` that we had so far and that I kept in this PR with `AllowDeath` in the `burn` impl

**Comment by stechu** [comment]
My general feeling is that `Keep_Alive` and `Allow_Death` is a bit adhoc now. We probably need to write down the correct semantics for when and when not to `Keep_Alive` in the context of both XCM and MantaPay.

**Comment by ghzlatarev** [comment]
> My general feeling is that `Keep_Alive` and `Allow_Death` is a bit adhoc now. We probably need to write down the correct semantics for when and when not to `Keep_Alive` in the context of both XCM and MantaPay.

In the context of XCM we should use `Allow_Death` for withdrawals/transfers imo. First, it's not any different than regular balance transfers, where this is allowed. Obviously if something relies on the account being alive, it will not be allowed to die anyway. I don't see any benefit in supporting both options besides UX> Second, the `can_decrease` impl that we rely on allows death, which means also the old transactors logic use `allow-death` by default. We can't make it a user input from xcm perspective because this is not an input to xcm instructions, or xtokens calls, etc.

In the context of MantaPay  we could make it a user input though. We'll have to slightly refactor the `FungibleLedger` trait and some of the manta-pay extrinsics, but I'm not sure if it's really needed.

@bhgomes

**Comment by stechu** [comment]
Thanks for the PR.

File: primitives/src/xcm.rs
Comment (stechu) [inline primitives/src/xcm.rs]: Rust doc about the semantics of this function.

**Comment by stechu** [comment]
A general comment on the design:
we need a clear documentation on the `allowDeath` behavior of all methods defined inside `FungibleLedger`, like 
`transfer`, `burn` etc. Also since `FungibleLedger` is used in both `mantaPay` and cross-chain token transfer. We need to examine in each use case, what is the `allowDeath` situation that makes sense.

**Comment by Garandor** [comment]
merge conflicts to resolve

**Review by Garandor — COMMENTED** [review]
LGTM, as it's an XCM PR I'd like to ask @Dengjianping and @stechu to review for approval here as well

File: pallets/asset-manager/src/lib.rs
Comment (Dengjianping) [inline pallets/asset-manager/src/lib.rs]: Seems, it's just for testing,  but will be compiled into native binary.
Is it possible to add one more attribute here `#[cfg(test)]`?
But I'm not sure it would compile.

File: primitives/src/xcm.rs
Comment (Dengjianping) [inline primitives/src/xcm.rs]: `trace` might print too much redundant logs, `debug` would be more friendly, imo.

File: primitives/src/xcm.rs
Comment (Dengjianping) [inline primitives/src/xcm.rs]: I suppose `result::Result::` can be removed?

File: primitives/src/assets.rs
Comment (Dengjianping) [inline primitives/src/assets.rs:494]: For native assets, it's transferred instead of burned, kind of confused to me. Perhaps leaving some comments here would make more sense.

File: runtime/calamari/Cargo.toml
Comment (Dengjianping) [inline runtime/calamari/Cargo.toml]: Where is this crate used? I mean I didn't find any type or trait is import in calamari-runtime.

File: primitives/src/xcm.rs
Comment (Dengjianping) [inline primitives/src/xcm.rs:418]: Does it work for account type like `AccountKey20`?

**Review by Dengjianping — CHANGES_REQUESTED** [review]

File: primitives/src/assets.rs
Comment (ghzlatarev) [reply primitives/src/assets.rs:494]: It is not transferred, the funciton is `withdraw`. The `Transfer` parameter are `reasons` which are used for `frozen()` checks

File: runtime/calamari/Cargo.toml
Comment (ghzlatarev) [reply runtime/calamari/Cargo.toml]: This is not used actually, will remove.

File: primitives/src/xcm.rs
Comment (ghzlatarev) [reply primitives/src/xcm.rs:418]: Depends on the runtime configuration. In our particular case we don't handle `AccountKey20` locations. But remember that the expected sender origin is actually the parachain sovereign account, not a user account. That's why we don't need to handle this conversion.
User account origin logic is handled on the sender side. Once the message is passed to the backend transportation system, the origin is changed to the parachain id of the sender.

File: pallets/asset-manager/src/lib.rs
Comment (ghzlatarev) [reply pallets/asset-manager/src/lib.rs]: It doesn't compile, can't find the function when running tests.

Comment (ghzlatarev) [reply pallets/asset-manager/src/lib.rs]: This is for convenience while testing, I don't think there's any risk to it.
But the other option is to set the proper native and non native start at genesis.
What do you guys think @Dengjianping @stechu @Garandor

**Comment by Dengjianping** [comment]
Please resolve clippy warnings.

**Review by Dengjianping — DISMISSED** [review]
lgtm!
Great PR!

File: CHANGELOG.md
Comment (Garandor) [inline CHANGELOG.md]: I dont think you should be deleting v3.2.0 changelog entries here

**Review by Garandor — CHANGES_REQUESTED** [review]
Forgot i never sent this review 🥇

File: CHANGELOG.md
Comment (ghzlatarev) [reply CHANGELOG.md]: oh my bad this must've been a bad merge conflict resolution

**Review by Dengjianping — APPROVED** [review]
lgtm!

**Review by Garandor — APPROVED** [review]
LGTM if you fix the changelog in #614 after merging this

**Pull request merged**
